text/0000-modular-crypto-service.md

```diff
@@ -0,0 +1,164 @@
+---
+layout: default
+title: RFC Template
+nav_order: 3
+---
+
+- Feature Name: modular_crypto_service
+- Start Date: 2020-09-19
+- RFC PR: (leave this empty)
+- Fabric Component: core, fabric-sdk-*, fabric-ca
+- Fabric Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+This RFC aims to provide broader crypto service configurability or pluggable capability. 
+It mainly changes design of crypt/X509 usage and hash algorithm opt-in in several layer.  
+
+# Motivation
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected
+outcome?
+
+Following is the original initiative post, translated from Chinese.
+```
+The Hyperledger Fabric code project has a very wide range of adoptions in many domestic industries and institutions, and has a very basic role in the application of the blockchain industry.
+According to Chinese policy requirements, it is necessary to support the national secret (GM) encryption algorithm. Many domestic enterprises and Fabric enthusiasts have successively carried out some corresponding national secret (GM) fabric forks with outstanding achievements. However, in this matter, there is serious duplication of working hours and waste of human resources.
+Here, the Hyperledger Technical Working Group China (TWGC) proposes that the majority of Fabric enthusiasts and users gather and work together to create a standard national cryptographic project of Hyperledger Fabric, which is shared by everyone, and finally meet the needs of community maintainer request, merge into the Fabric up-stream.
+```
+We now have 3 streams of China crypto libraries under Hyperledger-TWGC github organization. 
+Some of them has also provided successful ane enterprise-proven bccsp implementations. 
+But as along as we go deeper into fabric source, we notice implementing bccsp only could not fulfill all of Chinese national crypto specification requirements.
+There are some cypro library usages outside of bccsp, such as communication protocol, X509 format conversion. 
```

Comment:
The Fabric core is very, very tightly coupled with x509 standard:

All communication assumes TLS which uses x509 and doesn't have pluggable algorithms (though, it's possible to fork the TLS library and make your own version of it).

Fabric authentication infrastructure uses the BCCSP MSP which is also heavily based on x509 and doesn't abstract it out in any way. 

It's going to be a lot of work to overhaul the entire code base to abstract out x509, and, if we do this, it'll make backporting changes and bug fixes a pain because the code base would diverge too much.